### DISCUSSION TITLE
How to record extension host performance in code-server?

### DISCUSSION THREAD

**Discussion by chouzz** [coder/code-server#6823] [Q&A] [answered]
In the desktop version of VSCode, you can record the extension host process to analyze performance issues via the context menu. This is really useful for identifying low-performance extensions or functions based on the results. However, in code-server, there isn't a similar command, and even when trying to record performance in Chrome DevTools, it seems that it doesn't capture useful information like the desktop version does.

- context menu to record extension host performance in vscode desktop version.
![image](https://github.com/coder/code-server/assets/18023066/d539d1f2-f31f-4090-93e1-4401c05c9fb0)

- You can found performance issue related to `spawnSync` or `execSync`.
![image](https://github.com/coder/code-server/assets/18023066/7f04c76b-1e36-46b2-83f5-7f535187790a)

- But if you tried record performance in `code-server`, it's very different compared to the command. In fact, you can't find any useful information in the extension host workers; there are no functions that come from extensions.
![image](https://github.com/coder/code-server/assets/18023066/23192843-999e-4392-84cd-9c62d8e2722d)

**Comment by code-asher** [comment]
The web version of VS Code has an extension host that runs on the server and an extension host that runs in a web worker.  Only the web worker extension host would show up in Chrome's profiler, and only web extensions can run there (and most extensions are not web extensions, I believe).

As for why that context menu option is not available, you might have better luck asking upstream in https://github.com/microsoft/vscode about this feature for the web version.  I can only vaguely guess that this is something that worked in Electron that they have not been able to port to Node or web yet.

**Reply by chouzz** [reply]
So, in code-server, the extension host actually running in Linux, it's why Chrome Devtools can't record it. BTW, do you know of any other ways to analyze performance issues in extension host processes? @code-asher

**Reply by code-asher** [reply, marked as answer]
Yup exactly right.

I think it would be possible to use the Node profiler.  For example I just tried:

```
$ node --prof /path/to/code-server --auth none
```

Then you can check the processes and find the pid of the extension host.  In my case, 71247.

```
$ ps -eFjH | grep node
coder      69704    4075   69704    4075  0 200059 69424  9 20:42 pts/0    00:00:02       node --prof /home/coder/.local/lib/code-server-4.22.1 --port 3000
coder      69757   69704   69704    4075  0 289566 126844 2 20:42 pts/0    00:00:26         /nix/store/zsdfz83ilpz26gm44r0fh7h5bqabc4cf-nodejs-18.20.2/bin/node --prof /home/coder/.local/lib/code-server-4.22.1/out/node/entry
coder      71050   69757   69704    4075  0 231682 59456 20 20:43 pts/0    00:00:06           /nix/store/zsdfz83ilpz26gm44r0fh7h5bqabc4cf-nodejs-18.20.2/bin/node --prof /home/coder/.local/lib/code-server-4.22.1/lib/vscode/out/bootstrap-fork --type=fileWatcher
coder      71247   69757   69704    4075  0 2891821 143640 19 20:43 pts/0  00:00:17           /nix/store/zsdfz83ilpz26gm44r0fh7h5bqabc4cf-nodejs-18.20.2/bin/node --dns-result-order=ipv4first --prof /home/coder/.local/lib/code-server-4.22.1/lib/vscode/out/bootstrap-fork --type=extensionHost --transformURIs --useHostProxy=false
coder      75121   69757   69704    4075  0 181567 57876  1 20:45 pts/0    00:00:01           /nix/store/zsdfz83ilpz26gm44r0fh7h5bqabc4cf-nodejs-18.20.2/bin/node --prof /home/coder/.local/lib/code-server-4.22.1/lib/vscode/out/bootstrap-fork --type=ptyHost --logsPath /home/coder/.local/share/code-server/logs/20240606T204302
```

And then process the isolate file:

```
node --prof-process isolate-*-71247-v8.log
```

I am not sure how easy it is to use the results though.
